packets.go

```diff
@@ -259,11 +259,9 @@ func (mc *mysqlConn) writeAuthPacket(cipher []byte) error {
 	}
 
 	// Calculate packet length and get buffer with that size
-	data := mc.buf.takeSmallBuffer(pktLen + 4)
-	if data == nil {
-		// can not take the buffer. Something must be wrong with the connection
-		errLog.Print(ErrBusyBuffer)
-		return driver.ErrBadConn
+	data, err := mc.buf.takeSmallBuffer(pktLen + 4)
+	if err != nil {
+		return err
```

Comment:
Had some time to sort things out, I think that's correct behaviour to return `ErrBusyBuffer` because when we cannot take it then something must be wrong in the code (multiple functions are using the buffer at the same time) or in this case buffer has some unread data.

And only two functions `writePacket` and `readPacket` should return `ErrBadConn` since they do io.

So we should return `ErrBusyBuffer` anyway not `ErrBadConn`, and probably add additional checks to `Query` and `Exec` to return `ErrUnreadRows` to be more descriptive.

What do you think?